Feat: Create Select component

kanta1207 opened the pull request:
<!--

Title Naming Conventions:

- Start with a capitalized prefix followed by a imperative form of a verb just like a commit message.
- Refer to the "Commit Convention" for the prefix:
  https://github.com/genesis-tech-tribe/nishiki-frontend/blob/develop/docs/CONTRIBUTING.md#commit-convention.

Example:
- Fix: Resolve memory leak in data processing module
- Feat: Add two-factor authentication
- Refactor: Improve state management in React components
- Docs: Add Windows setup instructions

-->

## Overview

<!-- Provide a brief description of the changes introduced by this PR. -->
- Implemented reusable "Select" component.

## Review points

<!-- List the points you would like to be reviewed in detail and the points you are not confident about. -->
<!-- Delete this section if not needed -->

- Is the implementation of "Select" related components all good? I'll leave some comments & concerns as review comments 

## Screen Captures

<!-- Add screenshots or screen recordings of the changes -->
<!-- Necessary for UI changes -->
<!-- Delete this section if not needed -->


https://github.com/genesis-tech-tribe/nishiki-frontend/assets/99339182/e43556b6-ca50-4c05-9c01-1037b5b7bc6f



## Note

<!-- Write any note or comment. You can share your thoughts or ideas. -->
<!-- Delete this section if not needed -->

- I created "Select Trigger" component separately for now, since the logic for it is a bit long, and this is the only one component using "variants" and icon, of all "Select" related component. If there won't be any additional variants in the future, we can merge  "Select Trigger" into "Select" file.

## Assignee Checklist:

<!-- You can tick the checkboxes if not applicable -->

- [x] The base branch is correct (See: [Types of Branches](https://github.com/genesis-tech-tribe/nishiki-frontend/blob/develop/docs/CONTRIBUTING.md#types-of-branches))
- [x] The branch name follows the [Branch Naming Conventions](https://github.com/genesis-tech-tribe/nishiki-frontend/blob/develop/docs/CONTRIBUTING.md#branch-naming-conventions)
- [x] The correct assignees and reviewers have been designated for this PR
- [x] The coding style follows the [Coding Style Guide](https://github.com/genesis-tech-tribe/nishiki-frontend/blob/develop/docs/STYLEGUIDE.md)
- [x] All the related issues are associated with this PR
- [x] All criteria in the associated issues are met (please tick the checkboxes)
- [x] My changes do not generate new warnings or errors (especially in the console)

## Reviewer Checklist:

<!-- You can tick the checkboxes if not applicable -->

- [x] The code follows the generic best practices and our [Coding Style Guide](https://github.com/genesis-tech-tribe/nishiki-frontend/blob/develop/docs/STYLEGUIDE.md)
- [x] Code is well-commented and easy to understand
- [ ] UI changes accurately reflect the provided design

uskayyyyy commented on src/components/ui/Select/Select.tsx:
In my opinion, I think it would be better to use the position `item-aligned`. Because popping over the `<SelectContent>` below or above the `<SelectTrigger>` could make the `<SelectContent>` overflow the viewport, especially when there are many options.

https://github.com/genesis-tech-tribe/nishiki-frontend/assets/99148565/97a34f90-d6aa-4061-b2c2-fb11d44d88b7

If you stick to the position `popper`, you need to carefully adjust its size.
If you switch to the position `item-aligned`, you may also need to be aware that many properties are invalid with `item-aligned`.

uskayyyyy replied:
When the parent's y-paddings are removed, `<SelectContent>` shifts to the right and overflows the screen. Please take a look at the red part of the image. Could you inspect what's happening, please?

<img width="300" src="https://github.com/genesis-tech-tribe/nishiki-frontend/assets/99148565/b9872f96-603a-4563-bb0f-a5178c107793">

Additionally, I prefer to add some space between `<SelectTrigger>` and `<SelectContent>` although it's not defined in Figma. Please see the blue part of the image. What do you think of it?

uskayyyyy replied:
I have checked the appearance again, and it looks so nice now! Thank you!

uskayyyyy reviewed (changes requested):
Thank you for the excellent implementation! Please kindly take a look at the comments I left.

Overall, I think it might be good to ask the designers to design the `SelectContent` part.

kotaaaa reviewed (approved):
Nothing to add to @uskayyyyy 's comment, thank you for your great work!!

kanta1207 commented:
@kotaaaa 
Thank you for your review!!

@uskayyyyy 
Thank you for precise review and sorry for the delay.
I basically agreed with you, so I followed your review and fixed them, before replying to them.
 
 I added comment on how I fixed the points you mentioned, so please refer to them and take another review on this PR on your convenience! Thank you 🙌

uskayyyyy commented:
@kanta1207 
Thank you for updating!

> I added comment on how I fixed the points you mentioned

I don't see any comments from you...

uskayyyyy commented on src/components/ui/Select/Select.tsx:
```diff
@@ -0,0 +1,86 @@
+      'py-2 pl-6 pr-2',
+      'cursor-default select-none',
+      'data-[disabled]:pointer-events-none data-[disabled]:opacity-50',
+      className,
```
I tested how the `SelectContent` overflows on Safari of my iPhone SE (2nd gen). As you can see in the screen recording, the last option can still be overflown. Do you think this is acceptable?

https://github.com/genesis-tech-tribe/nishiki-frontend/assets/99148565/0b722047-b41f-4d11-8a31-c6058789985f

uskayyyyy commented:
@kanta1207 I have reviewed your update, and now I only want to discuss about this matter:

https://github.com/genesis-tech-tribe/nishiki-frontend/pull/114#discussion_r1433230281

kanta1207 commented:
> @kanta1207 Thank you for the excellent update!
> 
> > I added comment on how I fixed the points you mentioned
> 
> I don't see any comments from you though...

Sorry, I meant the comments on this PR on gitHub, not in code

kanta1207 commented:
> @kanta1207 I have reviewed your update, and now I only want to discuss about this matter:
> 
> [#114 (comment)](https://github.com/genesis-tech-tribe/nishiki-frontend/pull/114#discussion_r1433230281)

Thank you for all of the reviews!!

About  [#114 (comment)](https://github.com/genesis-tech-tribe/nishiki-frontend/pull/114#discussion_r1433230281), 
I thought I managed to fix it by adjusting the size, not making the position "item-aligned".

Do you want to  discuss about the number of height, or do you want to discuss about any other further concerns?

uskayyyyy commented:
@kanta1207
I knew that you adjusted the height of `SelectContent` though. On smaller screens, it still overflows.
Let's discuss it later verbally.

kanta1207 commented:
> @kanta1207 I knew that you adjusted the height of `SelectContent` though. On smaller screens, it still overflows. Let's discuss it later verbally.

After the discussion with you, I managed to fix the height (made it more flexible) by using [calc()](https://stackoverflow.com/questions/65976223/how-to-use-calc-in-tailwind-css)! Please take a look at them again 🙏

uskayyyyy reviewed (approved):
@kanta1207 
Thank you for your excellent work! I will leave it up to you to decide whether to implement the design of `SelectContent` in this PR or a separate ticket.
If you decide to handle it in this PR, please re-request me for a review.

kanta1207 commented:
> @kanta1207 Thank you for your excellent work! I will leave it up to you to decide whether to implement the design of `SelectContent` in this PR or a separate ticket. If you decide to handle it in this PR, please re-request me for a review.

Thank you for all the precise reviews! 
I'll merge this PR for now, and create another smaller issue(ticket) to apply the design from Figma to SelectContent
